clients/src/main/java/org/apache/kafka/common/network/PlaintextChannelBuilder.java

```diff
@@ -70,7 +81,9 @@ public void authenticate() throws IOException {}
         @Override
         public KafkaPrincipal principal() {
             InetAddress clientAddress = transportLayer.socketChannel().socket().getInetAddress();
-            return principalBuilder.build(new PlaintextAuthenticationContext(clientAddress));
+            // listenerName should only be null in Client mode
+            String listenerNameStr = listenerName != null ? listenerName.value() : null;
```

Comment:
This seems a bit brittle. Maybe it should be `Optional<ListenerName>`?

Reply:
We should never be calling `principal()` in client mode.

Reply:
We should probably not have the null check then, better fail fast. Same for the other case.